ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,777 @@
+  }
+
+  /**
+   * Return the {@link MessageFormatFlags} to associate with a getBlob operation.
+   * @return {@link MessageFormatFlags#Blob}
+   */
+  @Override
+  MessageFormatFlags getOperationFlag() {
+    return MessageFormatFlags.Blob;
+  }
+
+  /**
+   * Do all that needs to be done (cleanup, notification, etc.) on chunk completion and mark the state of the chunk
+   * appropriately.
+   * @param chunk the chunk that has completed.
+   */
+  private void onChunkOperationComplete(GetChunk chunk) {
+    if (chunk.getChunkException() != null) {
+      // if operation callback was already called, then this exception will have to be notified as part of the
+      // read callback.
+      setOperationException(chunk.getChunkException());
+    }
+    if (chunk == firstChunk) {
+      if (operationCallbackInvoked.compareAndSet(false, true)) {
+        // Complete the operation from the caller's perspective, so that the caller can start reading from the
+        // channel. The operation will not be marked as complete internally as subsequent chunk retrievals and
+        // channel writes will need to happen and for that, this operation needs the GetManager to poll it
+        // periodically. If any exception is encountered while processing subsequent chunks, those will be notified
+        // during the channel read.
+        ReadableStreamChannel result = getOperationException() == null ? this : null;
+        operationCompleteCallback
+            .completeOperation(operationFuture, operationCallback, result, getOperationException());
+      }
+    }
+    chunk.postCompletionCleanup();
+    maybeWriteToChannel();
+  }
+
+  /**
+   * Handle the given {@link ResponseInfo} by handing it over to the appropriate chunk that issued the request.
+   * @param responseInfo the {@link ResponseInfo} to be handled.
+   */
+  @Override
+  void handleResponse(ResponseInfo responseInfo) {
+    GetChunk getChunk =
+        correlationIdToGetChunk.remove(((RequestOrResponse) responseInfo.getRequest()).getCorrelationId());
+    getChunk.handleResponse(responseInfo);
+    if (getChunk.isComplete()) {
+      onChunkOperationComplete(getChunk);
+    }
+  }
+
+  /**
+   * Attempt to write the data associated with the blob to the channel passed in by the caller (if the caller has
+   * done so).
+   */
+  private void maybeWriteToChannel() {
+    // if there are chunks available to be written out, do now.
+    if (firstChunk.isComplete() && readCalled) {
+      while (indexOfNextChunkToWriteOut < numChunksTotal && chunkIndexToBuffer
+          .containsKey(indexOfNextChunkToWriteOut)) {
+        ByteBuffer chunkBuf = chunkIndexToBuffer.remove(indexOfNextChunkToWriteOut);
+        asyncWritableChannel.write(chunkBuf, chunkAsyncWriteCallback);
+        indexOfNextChunkToWriteOut++;
+      }
+      if (operationException.get() != null || numChunksWrittenOut.get() == numChunksTotal) {
+        completeRead();
+        operationCompleted = true;
```

Comment:
Will do. Though note that that is not the only scenario when `operationCompleted` will bet set to true.